File: 백준/Silver/2529. 부등호/부등호.py
import sys


def dfs1(depth):
    global max_num
    global flag1
    if depth == (n + 1):
        for i in range(len(opers)):
            oper = opers[i]
            if oper == '<' and array[i] < array[i + 1]:
                continue
            elif oper == '>' and array[i] > array[i + 1]:
                continue
            else:
                break
        else:
            # print(''.join(map(str,array)))
            max_num = max(max_num, ''.join(map(str, array)))
            flag1 = True
        return

    for i in range(9, -1, -1):
        if not visited1[i] and not flag1:
            visited1[i] = True
            array[depth] = i
            dfs1(depth + 1)
            visited1[i] = False


def dfs(depth):
    global min_num
    global flag
    if depth == (n + 1):
        for i in range(len(opers)):
            oper = opers[i]
            if oper == '<' and array[i] < array[i + 1]:
                continue
            elif oper == '>' and array[i] > array[i + 1]:
                continue
            else:
                break
        else:
            # print(''.join(map(str,array)))
            min_num = min(min_num, ''.join(map(str, array)))
            flag = True
        return

    for i in range(0, 10):
        if not visited[i] and not flag:
            visited[i] = True
            array[depth] = i
            dfs(depth + 1)
            visited[i] = False


n = int(sys.stdin.readline())

array = [0] * (n + 1)
opers = list(sys.stdin.readline().rstrip().split())
visited = [False] * 10
max_num = '0'
min_num = '9999999999'

flag = False
dfs(0)

flag1 = False
visited1 = [False] * 10
dfs1(0)
print(max_num)
print(min_num)

File: 백준/Silver/2529. 부등호/부등호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;


public class Main {

    static char[] op;
    static int[] arr;
    static int[] numbers = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
    static boolean[] visited = new boolean[10];
    static String min = "9999999999";
    static String max = "0";

    static void dfs(int depth) {
        if(depth > arr.length){
            return;
        }
        if (depth == arr.length) {
            for (int i = 0; i < op.length; i++) {
                if (op[i] == '>'){
                    if (arr[i] <= arr[i+1]){
                        return;
                    }
                }else if (op[i] == '<'){
                    if (arr[i] >= arr[i+1]){
                        return;
                    }
                }
            }

            String temp = "";
            for (int i : arr) {
                temp += i;
            }

            if (Long.parseLong(temp) < Long.parseLong(min)){
                min = temp;
            }

            if(Long.parseLong(temp) > Long.parseLong(max)){
                max = temp;
            }
            return;
        }

        for (int i = 0; i <= 9; i++) {
            if (!visited[i]) {
                visited[i] = true;
                arr[depth] = numbers[i];
                dfs(depth + 1);
                visited[i] = false;
            }

        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        op = new char[n];
        StringTokenizer st = new StringTokenizer(br.readLine());

        for (int i = 0; i < n; i++) {
            op[i] = st.nextToken().charAt(0);
        }
        arr = new int[n + 1];
        dfs(0);

        System.out.println(max);
        System.out.println(min);
    }
}
